prov/gni/src/gnix_nic.c

```diff
@@ -65,137 +65,121 @@ __desc_lkup_by_id(struct gnix_nic *nic, int desc_id)
- * Want to avoid using recursive locks.
- */
-static inline int __process_smsg(struct gnix_vc *vc)
+static int __nic_rx_overrun(struct gnix_nic *nic)
 {
-	int ret = FI_SUCCESS;
-	int status = GNI_RC_SUCCESS;
-	struct gnix_nic *nic = vc->ep->nic;
-	uint8_t tag = GNI_SMSG_ANY_TAG;
-	void *msg_ptr;
+	int i, max_id, ret;
+	struct gnix_vc *vc;
+	gni_return_t status;
+	gni_cq_entry_t cqe;
 
+	/* clear out the CQ */
+	while ((status = GNI_CqGetEvent(nic->rx_cq, &cqe)) == GNI_RC_SUCCESS);
+	assert(status == GNI_RC_NOT_DONE);
+
+	fastlock_acquire(&nic->vc_id_lock);
+	max_id = nic->vc_id_table_count;
+	fastlock_release(&nic->vc_id_lock);
 	/*
-	 * the nic lock is already held when this
-	 * function is invoked
+	 * TODO: optimization would
+	 * be to keep track of last time
+	 * this happened and where smsg msgs.
+	 * were found.
 	 */
-
-	do {
-		status = GNI_SmsgGetNextWTag(vc->gni_ep,
-					     &msg_ptr,
-					     &tag);
-		if (status == GNI_RC_SUCCESS) {
-			ret = nic->smsg_callbacks[tag](vc,
-						       msg_ptr);
+	for (i = 0; i < max_id; i++) {
+		ret = _gnix_test_bit(&nic->vc_id_bitmap, i);
+		if (ret) {
+			vc = __gnix_nic_elem_by_rem_id(nic, i);
+			ret = _gnix_vc_schedule(vc);
 			assert(ret == FI_SUCCESS);
 		}
-	} while (status == GNI_RC_SUCCESS);
+	}
 
-	if (status != GNI_RC_NOT_DONE) {
-		GNIX_WARN(FI_LOG_EP_DATA,
-			  "GNI_SmsgGetNextWTag returned %s\n",
-			  gni_err_str[status]);
-		ret = gnixu_to_fi_errno(status);
+	return FI_SUCCESS;
+}
+
+static int process_rx_cqe(struct gnix_nic *nic, gni_cq_entry_t cqe)
+{
+	int ret = FI_SUCCESS, vc_id = 0;
+	struct gnix_vc *vc;
+
+	vc_id =  GNI_CQ_GET_INST_ID(cqe);
+	vc = __gnix_nic_elem_by_rem_id(nic, vc_id);
+
+#if 1 /* Process RX inline with arrival of an RX CQE. */
+	if (unlikely(vc->conn_state != GNIX_VC_CONNECTED)) {
+		GNIX_INFO(FI_LOG_EP_CTRL,
+			  "Scheduling VC for RX processing (%p)\n",
+			  vc);
+		_gnix_set_bit(&vc->flags, GNIX_VC_FLAG_RX_PENDING);
+		ret = _gnix_vc_schedule(vc);
+		assert(ret == FI_SUCCESS);
+	} else {
+		GNIX_INFO(FI_LOG_EP_CTRL,
+			  "Processing VC RX (%p)\n",
+			  vc);
+		ret = _gnix_vc_dequeue_smsg(vc);
+		if (ret != FI_SUCCESS) {
+			GNIX_WARN(FI_LOG_EP_CTRL,
+					"_gnix_vc_dqueue_smsg returned %d\n",
+					ret);
+		}
 	}
+#else /* Defer RX processing until after the RX CQ is cleared. */
+	_gnix_set_bit(&vc->flags, GNIX_VC_FLAG_RX_PENDING);
+	ret = _gnix_vc_schedule(vc);
+	assert(ret == FI_SUCCESS);
+#endif
 
 	return ret;
 }
 
 static int __nic_rx_progress(struct gnix_nic *nic)
 {
-	int i, ret = FI_SUCCESS;
-	int vc_id = 0;
-	int max_id;
-	gni_return_t  status = GNI_RC_NOT_DONE;
+	int ret = FI_SUCCESS;
+	gni_return_t status = GNI_RC_NOT_DONE;
 	gni_cq_entry_t cqe;
-	struct gnix_vc *vc;
 
-	/*
-	 * first see if there are any CQE's on the
-	 * tx cq. Calling GNI_CqTestEvent does not require
-	 * getting the nic lock
-	 */
-
-#if 0
 	status = GNI_CqTestEvent(nic->rx_cq);
-	if (status == GNI_RC_NOT_DONE)
+	if (likely(status == GNI_RC_NOT_DONE))
```

Comment:
When fast polling fi_cq_read(), the common case will be not having a CQE ready.  I'm not sure this is positive for performance so I'll remove it.